Adding custom transport to Logger library

supratiksen opened the pull request:
Tested with translator (adhoc changes, not sending out PR) as well newly added unit tests. npm run lint shows some issues but most of them are unrelated to this PR. 

Since we now have 2 optional parameters, "undefined" is logged for the 2nd field (logObject), in case there is a 3rd field (loggingMetadata) in each of the logging statements. As an example I updated the warn method to avoid logging undefined, but I am not sure if we want to avoid logging it, or if we do, if there is a better solution?

supratiksen committed:
Adding custom transport

supratiksen committed:
Example update of warn method to show we can avoid logging 'undefined'

supratiksen commented:
@Shalinij is added to the review. #Closed

supratiksen commented:
@kurstory is added to the review. #Closed

supratiksen commented:
@shalinijoshi19 is added to the review. #Closed

arjun-msft commented on node/lib/Logger.ts:
```diff
@@ -20,62 +22,89 @@ export class Logger implements ILogger {
+    public removeTransport(transportObject: any): void {
+        this.logger.remove(transportObject);
+
+        this.transportList.pop();
```
how do you know pop pops the right transportObject off? #Closed

supratiksen replied:
fixed

---
In reply to: [105339784](https://github.com/openT2T/opent2t/pull/38#discussion_r105339784) [](ancestors = 105339784)

arjun-msft commented on node/lib/Logger.ts:
```diff
@@ -20,62 +22,89 @@ export class Logger implements ILogger {
                 level: this.globalLogLevel,
             });
 
             this.logger.configure({
```
wouldn't it be cleaner to configure the logger with this.transportList once rather than inside these two ifs? #Resolved

supratiksen replied:
we probably want to be able to configure the transports differently (logging levels, etc.)

---
In reply to: [105340151](https://github.com/openT2T/opent2t/pull/38#discussion_r105340151) [](ancestors = 105340151)

arjun-msft commented on node/lib/Logger.ts:
```diff
@@ -20,62 +22,89 @@ export class Logger implements ILogger {
         }
 
-        let consoleTransport = new winston.transports.Console({
+        this.consoleTransport = new winston.transports.Console({
```
we should have a way to turn this off too #Resolved

shalinijoshi19 replied:
I agree -- now that we have custom transports being configured. 

---
In reply to: [105340367](https://github.com/openT2T/opent2t/pull/38#discussion_r105340367) [](ancestors = 105340367)

arjun-msft commented on node/lib/Logger.ts:
```diff
@@ -20,62 +22,89 @@ export class Logger implements ILogger {
 
-        if (Logger.flag === true && filename ) {
-            let fileTransport = new winston.transports.File({
+        if (Logger.flag === true && filename) {
```
what is Logger.flag? #Resolved

supratiksen replied:
to prevent a transport being added twice, remember the error we saw the 1st time I demo-ed this, this was to fix that. Of course we have moved from .add to .configure, so we might not need that check anymore..

---
In reply to: [105340423](https://github.com/openT2T/opent2t/pull/38#discussion_r105340423) [](ancestors = 105340423)

arjun-msft commented on node/lib/Logger.ts:
```diff
@@ -20,62 +22,89 @@ export class Logger implements ILogger {
+    public warn(msg: string, logObject?: any, loggingMetadata?: string): void {
+       if(typeof(logObject === 'undefined') && typeof(loggingMetadata) !== 'undefined')
+       {
+           this.logger.error(msg, loggingMetadata);
```
why not warn? #Resolved

supratiksen replied:
typo, was updating the error method 1st. Fixed

---
In reply to: [105340523](https://github.com/openT2T/opent2t/pull/38#discussion_r105340523) [](ancestors = 105340523)

supratiksen committed:
minor fixes, and addressing some of arjunl's feedback

supratiksen committed:
removing exact transport instead of last one pushed in

shalinijoshi19 commented on node/lib/ILogger.ts:
```diff
@@ -1,8 +1,11 @@
-    verbose(msg: string, logObject?: any): void;
-    debug(msg: string, logObject?: any): void;
+    // Example of loggingMetadata is correlationVector
+    error(msg: string, logObject?: any, loggingMetadata?: string): void; 
```
>  error(msg: string, logObject?: any, loggingMetadata?: string): void;  [](start = 2, length = 71)

How does this work usage-wise, if i have no logObject but loggingMetadata? error("myMessage", "my-CoVector") #Closed

supratiksen replied:
example in opent2ttests.ts

---
In reply to: [105465737](https://github.com/openT2T/opent2t/pull/38#discussion_r105465737) [](ancestors = 105465737)

shalinijoshi19 commented on node/test/OpenT2TTests.ts:
```diff
@@ -141,12 +142,29 @@ test("Logger with default parameters can be instantiated multiple times", async
     t.is(logger.getConfiguredTransports().length, 2);
 });
 
+test("Logger with default parameters can log custom loggingMetadata parameter like a correlationVector", async t => {
```
nit: horizontal readability - split across two lines #Resolved

supratiksen replied:
Fixed

---
In reply to: [105495418](https://github.com/openT2T/opent2t/pull/38#discussion_r105495418) [](ancestors = 105495418)

shalinijoshi19 commented on node/test/OpenT2TTests.ts:
```diff
@@ -141,12 +142,29 @@ test("Logger with default parameters can be instantiated multiple times", async
+    let logger = new Logger();
+    let stringObject = "Random String"
+    logger.info("Writing default level to default console with loggingMetadata:", stringObject, "ABC123");
+    logger.warn("writing warn level to default console with loggingMetadata:", undefined, "XYZ456");
```
this would be not-user-friendly - might want to switch the parameters around or as Arjun suggested better still, make this a dictionary where the first is the metadata.

shalinijoshi19 commented:
:clock1:

supratiksen commented on node/lib/Logger.ts:
```diff
@@ -20,62 +22,89 @@ export class Logger implements ILogger {
         }
 
-        let consoleTransport = new winston.transports.Console({
+        this.consoleTransport = new winston.transports.Console({
```
We can turn it off using removeTransport(winston.transports.Console);

---
In reply to: [105495262](https://github.com/openT2T/opent2t/pull/38#discussion_r105495262) [](ancestors = 105495262,105340367)

supratiksen commented:
> const  testLogFileName = "myloggertest.log";

hey

---
Refers to: node/test/OpenT2TTests.ts:23 in 95e06a8. [](commit_id = 95e06a8407b1712e0bc8224b53332ff20299b94e, deletion_comment = False)

supratiksen committed:
Switched logObject & logMetadata. Commented out examples of using Map or
Associative Array

shalinijoshi19 commented on node/lib/ILogger.ts:
```diff
@@ -1,8 +1,15 @@
+    warn(msg: string, logMetadata?: string, logObject?: any): void;
+/*    warn(msg: string, logObject?: any, logMetadata?: string): void;
+    warn(msg: string, logParams?: Map<any, string>): void;
+    warn(msg: string, logParams?: Array<any>): void;*/
```
Why is this commented out? Remove the comment or explain a comment explaining the comment if the commented out code is really needed. #Resolved

shalinijoshi19 commented on node/lib/Logger.ts:
```diff
@@ -52,30 +55,64 @@ export class Logger implements ILogger {
 
-    public warn(msg: string, logObject?: any): void {
-        this.logger.warn(msg, logObject);
+/*    public warn(msg: string, logObject?: any, logMetadata?: string): void {
```
Again why is this commented out?  #Resolved

shalinijoshi19 commented on node/test/OpenT2TTests.ts:
```diff
@@ -141,12 +142,45 @@ test("Logger with default parameters can be instantiated multiple times", async
+/*    var temp:any = new Array();
+    temp["logObject"] = stringObject;
+    temp["logMetadata"] = "ABC123";
+    logger.warn("writing warn level:", temp);*/
```
Why are you commenting out code all over the place?

shalinijoshi19 commented:
:clock1:

shalinijoshi19 reviewed (changes requested):
Comments in Codeflow. My biggest gripe here is the presence of so many comments without explaining why the code is commented out  - if these are guides/samples have samples in the readme and yank ou thte commented out code

supratiksen committed:
removed commented code. swapped logObject & logMetadata for all logging
methods. Next up will be a PR with hubController changes in cloud repo
and updates to logging calls in translator repo

supratiksen committed:
Adding ability to change logLevel for any transport at runtime
